Site Editor: Implement a settings object filter

jeyip opened the pull request:
Potentially solves https://github.com/WordPress/gutenberg/issues/33736

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
The **post editor** has the `block_editor_settings_all` and `block_editor_settings` filter that allows us to modify the `settings` object injected into the block editor. A similar filter, however, doesn't exist for the site editor.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
**There should be no visible changes to the UI.**

1. Set up local Gutenberg dev environment
2. Activate a block based theme
3. Navigate to the site editor
4. Smoke test the site editor to ensure that updates still persist correctly, block insertion still functions as expected, etc.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jeyip committed:
Add block editor settings filter for site editor init

The post editor has the block_editor_settings_all and block_editor_settings filter that allows us to modify the settings object injected into the block editor. A similar filter, however, doesn't exist for the site editor. We apply the same filter here for the site editor.

jeyip commented on lib/full-site-editing/edit-site-page.php:
```diff
@@ -101,7 +101,7 @@ function gutenberg_edit_site_init( $hook ) {
 		)
 	);
-	$settings = gutenberg_experimental_global_styles_settings( $settings );
+	$settings = apply_filters( 'block_editor_settings_all', $settings );
```
I remove `gutenberg_experimental_global_styles_settings` because it is already being added as a callback to the `block_editor_settings_all` filter in `lib/global-styles.php`

https://github.com/WordPress/gutenberg/blob/1c733c07a4f0d48d5f594c0af44047ac11d5347a/lib/global-styles.php#L329-L334

mattwiebe replied:
We'll need to add a second param to the `apply_filters` call here, as it's called with `WP_Block_Editor_Context` in the second param [here in Core](https://github.com/WordPress/WordPress/blob/79685db062fa5efa5c31f133f177a0d58eef22b9/wp-includes/block-editor.php#L362-L370). 

It appears that, if we're not in a post editing context, we're just using `new WP_Block_Editor_Context()` with no params, like [here in the Widgets Block Editor](https://github.com/WordPress/gutenberg/blob/339a7d783ad53dc9d437b3b5764714843caea2b8/lib/widgets-page.php#L81). 

So sum up, we should have:

```php
$site_editor_context = new WP_Block_Editor_Context();
$settings = apply_filters( 'block_editor_settings_all', $settings, $site_editor_context );
```

Oh! Also calls to `apply_filters` are always documented. Since `block_editor_settings_all` is already documented in Core, referencing that is sufficient. Now we get:


```php
$site_editor_context = new WP_Block_Editor_Context();
/** This filter is documented in wp-includes/wp-block-editor.php */ 
$settings = apply_filters( 'block_editor_settings_all', $settings, $site_editor_context );
```

jeyip replied:
Awesome callouts. Thanks @mattwiebe

mattwiebe reviewed (changes requested):
Left a comment about needing another param and Core's filter documentation convention. Otherwise, this is great, and good call on not needing to call `gutenberg_experimental_global_styles_settings` directly, since it runs on the filter.

jeyip committed:
Add site editor context param to apply filters

We need to add a second param to our call to apply filters because it is included in Core WordPress.

jeyip committed:
Add comment for call to apply filters

gziolo commented on lib/full-site-editing/edit-site-page.php:
```diff
@@ -101,7 +101,10 @@ function gutenberg_edit_site_init( $hook ) {
 	);
-	$settings = gutenberg_experimental_global_styles_settings( $settings );
+
+	$site_editor_context = new WP_Block_Editor_Context();
```
I'm not quite sure why the edit site page doesn't integrate this filter like other screens. If we decide to change it, we should refactor the code to use `gutenberg_get_block_editor_settings` to handle it all behind the scenes rather that duplicating a similar logic.

jeyip replied:
Makes sense to me. I'll make the changes. 👍

jeyip replied:
Updated in 72b175c

jeyip committed:
Refactor with gutenberg_get_block_editor_settings

jeyip commented on lib/full-site-editing/edit-site-page.php:
```diff
@@ -89,19 +89,17 @@ function gutenberg_edit_site_init( $hook ) {
-			'__experimentalBlockPatterns'          => WP_Block_Patterns_Registry::get_instance()->get_all_registered(),
-			'__experimentalBlockPatternCategories' => WP_Block_Pattern_Categories_Registry::get_instance()->get_all_registered(),
-		)
+	$custom_settings     = array(
```
`$custom_settings` will be merged with `gutenberg_get_default_block_editor_settings` in the `gutenberg_get_block_editor_settings` function, which is why we remove the `array_merge` here.

 https://github.com/WordPress/gutenberg/blob/c96feebc4dafa087b5d247ad9c3a6ace5cb96d3c/lib/compat/wordpress-5.8/block-editor.php#L264-L272

jeyip commented on lib/full-site-editing/edit-site-page.php:
```diff
@@ -89,19 +89,17 @@ function gutenberg_edit_site_init( $hook ) {
 	$current_screen->is_block_editor( true );
 
-	$settings = array_merge(
-		gutenberg_get_default_block_editor_settings(),
```
`gutenberg_get_default_block_editor_settings` will be merged in the `gutenberg_get_block_editor_settings` function, which is why we remove it from here.

https://github.com/WordPress/gutenberg/blob/c96feebc4dafa087b5d247ad9c3a6ace5cb96d3c/lib/compat/wordpress-5.8/block-editor.php#L264-L272

gziolo reviewed (approved):
Code looks good and it aligns with configuration for other screens with the block editor.

jeyip commented:
Thanks @gziolo! Planning to let this PR sit a a few more days so that @mattwiebe and @youknowriad get a chance to take a look.

youknowriad commented:
If @gziolo is happy about multi editor settings, I'm happy too :)

mattwiebe reviewed (approved):
Looks good to me!
